### PR TITLE
Fix outdated behat scenarios

### PR DESCRIPTION
**Pull request by marekrzytki**

### CODE REVIEW

File: features/shop/paying_with_mollie_during_checkout.feature [hunk 1]
```diff
@@ -8,45 +8,64 @@ Feature: Paying with Mollie during checkout
...
     @ui
     Scenario: Cancelling the payment
+        Given the store has a product "PHP T-Shirt" priced at "$9.99"
         Given I added product "PHP T-Shirt" to the cart
         And I have proceeded selecting "Mollie" payment method
-        When I confirm my order with Mollie payment
-        And I cancel my Mollie payment
+        When I select "creditcard" as my payment option
+        Then I press "Next"
+        When I confirm my order
```
Comment (senghe) [inline features/shop/paying_with_mollie_during_checkout.feature:37]: When I confirm it gets cancelled? It doesn't sound good in my opinion

File: features/shop/paying_with_mollie_during_checkout.feature [hunk 2]
```diff
@@ -8,45 +8,64 @@ Feature: Paying with Mollie during checkout
...
+        Then I should be notified that my payment has been cancelled
+        And I should be able to pay again
+        And I fix payment amount to pay successfully
+        When I select "creditcard" as my payment option
+        Then I press "Pay"
         Then I should be notified that my payment has been completed
         And I should see the thank you page
 
     @ui
     Scenario: Retrying the payment and failing
```
Comment (senghe) [inline features/shop/paying_with_mollie_during_checkout.feature:58]:
```suggestion
    Scenario: Retrying the payment with failure
```

File: features/shop/paying_with_mollie_during_checkout.feature (hunk 1 above)
Comment (senghe) [reply features/shop/paying_with_mollie_during_checkout.feature:37]: Ah ok, I've got it. It's being cancelled from admin.

File: tests/Behat/Context/Setup/OrderContext.php [hunk 3]
```diff
@@ -54,26 +55,27 @@ public function thisOrderWithMolliePaymentIsAlreadyPaid(OrderInterface $order):
...
         foreach ($order->getPayments() as $payment) {
             /** @var PaymentMethodInterface $paymentMethod */
             $paymentMethod = $payment->getMethod();
 
-            if (true === in_array($paymentMethod->getGatewayConfig()->getFactoryName(), [MollieGatewayFactory::FACTORY_NAME, MollieSubscriptionGatewayFactory::FACTORY_NAME], true)) {
+            $gatewayConfig = $paymentMethod->getGatewayConfig();
+
+            Assert::notNull($gatewayConfig);
+
+            if (true === in_array($gatewayConfig->getFactoryName(), [MollieGatewayFactory::FACTORY_NAME, MollieSubscriptionGatewayFactory::FACTORY_NAME], true)) {
```
Comment (senghe) [inline tests/Behat/Context/Setup/OrderContext.php:68]: I would put it into private method for better readibility

File: tests/Behat/Context/Ui/Shop/ProductContext.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+     * @When I fix payment amount to pay successfully
+     */
+    public function iFixPaymentAmountToPaySuccessfully(): void
+    {
+        /** @var PaymentInterface $payment */
+        $payment = $this->paymentRepository->findOneBy([
+            'state' => PaymentInterface::STATE_NEW,
+        ]);
+
+        $payment->setAmount(1999);
```
Comment (senghe) [inline tests/Behat/Context/Ui/Shop/ProductContext.php:37]: I would put the 1999 and other values to a separate class and name them to know what they mean

File: tests/Behat/Page/Admin/PaymentMethod/CreatePage.php [hunk 5]
```diff
@@ -75,6 +63,7 @@ public function containsErrorWithMessage(string $message, bool $strict = true):
                 return true;
             }
 
+            /** @phpstan-ignore-next-line  */
```
Comment (senghe) [inline tests/Behat/Page/Admin/PaymentMethod/CreatePage.php:66]: Why we ignore it?

File: tests/Behat/Context/Ui/Shop/CheckoutContext.php [hunk 6]
```diff
@@ -14,124 +14,24 @@
...
-use Tests\BitBag\SyliusMolliePlugin\Behat\Mocker\MollieApiMocker;
-use Tests\BitBag\SyliusMolliePlugin\Behat\Page\External\PaymentPageInterface;
 use Tests\BitBag\SyliusMolliePlugin\Behat\Page\Shop\Checkout\CompletePageInterface;
 
 final class CheckoutContext extends RawMinkContext implements Context
 {
     /** @var CompletePageInterface */
     private $summaryPage;
 
-    /** @var ShowPageInterface */
```
Comment (senghe) [inline tests/Behat/Context/Ui/Shop/CheckoutContext.php:27]: Why so much code is being removed?

File: tests/Behat/Context/Setup/MollieContext.php [hunk 7]
```diff
@@ -133,12 +133,15 @@ public function theStoreHasAPaymentMethodWithACodeAndMollieSubscriptionPaymentGa
...
-        $paymentMethod->getGatewayConfig()->setConfig([
+        $gatewayConfig = $paymentMethod->getGatewayConfig();
+        Assert::notNull($gatewayConfig);
+        $gatewayConfig->setConfig([
             'api_key' => 'test',
             'payum.http_client' => '@bitbag_sylius_mollie_plugin.mollie_api_client',
             'api_key_test' => $this->mollieApiKeyTest,
             'profile_id' => $this->mollieProfileId,
             'environment' => null,
+            'loggerLevel' => 2,
```
Comment (senghe) [inline tests/Behat/Context/Setup/MollieContext.php:144]: What does exactly mean 2?

File: src/Validator/Constraints/PaymentMethodCheckoutValidator.php [hunk 8]
```diff
@@ -62,4 +63,13 @@ public function validate($value, Constraint $constraint): void
         }
         $this->context->buildViolation($constraint->message)->setTranslationDomain('messages')->addViolation();
     }
+
+    private function isMollieGatewayFactory(GatewayConfigInterface $gateway): bool
```
Comment (senghe) [inline src/Validator/Constraints/PaymentMethodCheckoutValidator.php:67]: I saw somewhere below same code but with no private method separation. How about creating trait for this?

**Review by senghe — CHANGES_REQUESTED** [review]

File: tests/Behat/Context/Ui/Shop/CheckoutContext.php (hunk 6 above)
Comment (marekrzytki) [reply tests/Behat/Context/Ui/Shop/CheckoutContext.php:27]: Functions were depending on outdated mockery

File: tests/Behat/Context/Setup/OrderContext.php (hunk 3 above)
Comment (marekrzytki) [reply tests/Behat/Context/Setup/OrderContext.php]: Created separate Checker - fixed

File: tests/Behat/Context/Ui/Shop/ProductContext.php (hunk 4 above)
Comment (marekrzytki) [reply tests/Behat/Context/Ui/Shop/ProductContext.php]: Moved to variable - fixed

File: tests/Behat/Context/Setup/MollieContext.php (hunk 7 above)
Comment (marekrzytki) [reply tests/Behat/Context/Setup/MollieContext.php]: Fixed

File: spec/Checker/Gateway/MollieGatewayFactoryCheckerSpec.php [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        ];
+
+        $gateway->getFactoryName()
+            ->willReturn('mollie_subscription');
+
+        $this->isMollieGateway($gateway)
+            ->shouldReturn(true);
+    }
+
+    public function it_returns_false_if_gateway_name_not_mollie(
```
Comment (senghe) [inline spec/Checker/Gateway/MollieGatewayFactoryCheckerSpec.php:59]:
```suggestion
    public function it_returns_false_if_gateway_name_is_not_mollie(
```

**Review by senghe — APPROVED** [review]

**Pull request merged**